tests/assessment/test_validity_checks.py

```diff
@@ -249,35 +258,269 @@ def test_unfurl_array_annotations(
+                "nhomalt_amr_XX": 5,
+                "AC": 10,
+                "AC_XX": 10,
+            },
+        },
+        {
+            "locus": hl.locus("chrY", 1000000, reference_genome="GRCh38"),
+            "info": {
+                "nhomalt": 5,
+                "nhomalt_XX": hl.missing(hl.tint32),
+                "nhomalt_amr": hl.missing(hl.tint32),
+                "nhomalt_amr_XX": hl.missing(hl.tint32),
+                "AC_XX": hl.missing(hl.tint32),
+                "AC": 6,
+            },
+        },
+        {
+            "locus": hl.locus("chrY", 2000000, reference_genome="GRCh38"),
+            "info": {
+                "nhomalt": 5,
+                "nhomalt_XX": 3,
+                "nhomalt_amr": hl.missing(hl.tint32),
+                "nhomalt_amr_XX": hl.missing(hl.tint32),
+                "AC_XX": hl.missing(hl.tint32),
+                "AC": 6,
+            },
+        },
+    ]
 
-    # Call the check_array_struct_missingness function.
-    missingness = check_array_struct_missingness(ht, indexed_array_annotations)
-
-    # Define the expected missingness percentages for each unfurled field.
-    # All 'adj' values have no missing data.
-    # AC_raw is missing only in row 4.
-    # AF_raw is missing in rows 2, 3, and 4.
-    # AN_raw is missing in all rows.
-    expected_missingness = {
-        "AC_adj": 0.0,
-        "AF_adj": 0.0,
-        "AC_raw": 0.25,
-        "AF_raw": 0.25,
-        "AN_eas_adj": 0.50,
-        "AN_eas_raw": 0.75,
-        "AN_sas_adj": 0.0,
-        "AN_sas_raw": 1.00,
-    }
+    ht = hl.Table.parallelize(
+        data,
+        hl.tstruct(
+            locus=hl.tlocus(reference_genome="GRCh38"),
+            info=hl.tstruct(
+                nhomalt=hl.tint32,
+                nhomalt_XX=hl.tint32,
+                nhomalt_amr=hl.tint32,
+                nhomalt_amr_XX=hl.tint32,
+                AC=hl.tint32,
+                AC_XX=hl.tint32,
+            ),
+        ),
+    )
+    ht = ht.key_by("locus")
+    return ht
 
-    # Validate each field's missingness percentage.
-    for field, expected_value in expected_missingness.items():
-        assert missingness[field] == expected_value, (
-            f"Mismatch in missingness for field '{field}': "
-            f"expected {expected_value}, got {missingness[field]}"
-        )
+
+def test_check_sex_chr_metrics_logs(ht_for_check_sex_chr_metrics) -> None:
+    """Test that check_sex_chr_metrics produces the expected log messages."""
+    ht = ht_for_check_sex_chr_metrics
+    info_metrics = [
+        "nhomalt",
+        "nhomalt_XX",
+        "nhomalt_amr",
+        "nhomalt_amr_XX",
+        "AC",
+        "AC_XX",
+    ]
+    contigs = ["chrX", "chrY"]
+    verbose = False
+
+    # Redirect logs to a buffer.
+    log_stream = StringIO()
+    logger = logging.getLogger("gnomad.assessment.validity_checks")
+    handler = logging.StreamHandler(log_stream)
+    logger.addHandler(handler)
+    logger.setLevel(logging.INFO)
```

Comment:
Have you looked into pytest's caplog? I think it could replace a lot of this and does exactly what youve set up here. It's built into pytest and since were already using it, it would reduce a lot of this code:

https://docs.pytest.org/en/stable/how-to/logging.html

Reply:
oh didn't know about that, thanks!